go-controller/pkg/ovn/controller/admin_network_policy/status.go

```diff
@@ -45,111 +46,127 @@ const (
-			Status:  metav1.ConditionFalse,
-			Reason:  policyNotReadyReason,
-			Message: message,
-		})
-		_, err = c.anpClientSet.PolicyV1alpha1().AdminNetworkPolicies().UpdateStatus(context.TODO(), canp, metav1.UpdateOptions{})
-		return err
-	})
+func (c *Controller) updateANPStatusToNotReady(anpName, message string) error {
+	notReadyCondition := metav1.Condition{
+		Type:    policyReadyStatusType + c.zone,
+		Status:  metav1.ConditionFalse,
+		Reason:  policyNotReadyReason,
+		Message: message,
+	}
+	err := c.updateANPZoneStatusCondition(notReadyCondition, anpName)
 	if err != nil {
-		return fmt.Errorf("unable update the status of ANP %s, err: %v", anp.Name, err)
+		return fmt.Errorf("unable update the status of ANP %s, err: %v", anpName, err)
 	}
 	klog.V(3).Infof("Patched the status of ANP %v with condition type %v/%v and reason %s/%s",
-		anp.Name, policyReadyStatusType+zone, metav1.ConditionFalse, policyNotReadyReason, message)
+		anpName, policyReadyStatusType+c.zone, metav1.ConditionFalse, policyNotReadyReason, message)
 	return nil
 }
 
-// updateBANPStatusToReady updates the status of the policy to reflect that it is ready
-// Each zone's ovnkube-controller will call this, hence let's update status using retryWithConflict
-func (c *Controller) updateBANPStatusToReady(banp *anpapi.BaselineAdminNetworkPolicy, zone string) error {
-	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
-		latestBANP, err := c.banpLister.Get(banp.Name)
-		if err != nil {
-			return fmt.Errorf("unable to fetch BANP %s, err: %v", banp.Name, err)
-		}
-		cbanp := latestBANP.DeepCopy()
-		meta.SetStatusCondition(&cbanp.Status.Conditions, metav1.Condition{
-			Type:    policyReadyStatusType + zone,
-			Status:  metav1.ConditionTrue,
-			Reason:  policyReadyReason,
-			Message: "Setting up OVN DB plumbing was successful",
-		})
-		_, err = c.anpClientSet.PolicyV1alpha1().BaselineAdminNetworkPolicies().UpdateStatus(context.TODO(), cbanp, metav1.UpdateOptions{})
+func (c *Controller) updateANPZoneStatusCondition(newCondition metav1.Condition, anpName string) error {
+	anp, err := c.anpLister.Get(anpName)
+	if err != nil {
 		return err
-	})
+	}
+	existingCondition := meta.FindStatusCondition(anp.Status.Conditions, newCondition.Type)
+	if existingCondition == nil {
+		newCondition.LastTransitionTime = metav1.NewTime(time.Now())
+	} else {
+		if existingCondition.Status != newCondition.Status {
+			existingCondition.Status = newCondition.Status
+			existingCondition.LastTransitionTime = metav1.NewTime(time.Now())
+		}
+		existingCondition.Reason = newCondition.Reason
+		existingCondition.Message = newCondition.Message
+		newCondition = *existingCondition
+	}
+	applyObj := anpapiapply.AdminNetworkPolicy(anpName).
+		WithStatus(anpapiapply.AdminNetworkPolicyStatus().WithConditions(newCondition))
+	_, err = c.anpClientSet.PolicyV1alpha1().AdminNetworkPolicies().
+		ApplyStatus(context.TODO(), applyObj, metav1.ApplyOptions{FieldManager: c.zone})
```

Comment:
it is usually recommended to also set `Force=true` for `ApplyOptions`

Reply:
great catch, I learnt something new :) https://pkg.go.dev/k8s.io/client-go/applyconfigurations#hdr-Controller_Support fixing this

Reply:
this is done, resolving comment.